### PR TITLE
Limb class

### PR DESCRIPTION
**Pull request by zacgong**
Created a LimbBase class between JointBase and RobotBase

- Fully abstract base class with pure virtual kinematics and helper functions
- Incorporated into RobotBase with AddLimb function
- Simple_Limb example for 1DOF limb with joint at origin and straight end effector
- Testing in robot_example still a work in progress, having trouble compiling

### CODE REVIEW

File: include/kodlab_mjbots_sdk/limb_base.h [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+         * @return std::vector<float>
+         */
+        std::vector<float> get_torques();
+
+        /**
+         * @brief Set the positions of each joint in the leg
+         *
+         * @param positions
+         */
+
```
Comment (ethanmusser) [inline include/kodlab_mjbots_sdk/limb_base.h:129]: Again, the Doxygen comment block should directly precede the function/object/namespace it is documenting.
```suggestion
```

File: include/kodlab_mjbots_sdk/limb_base.h [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        void set_positions(std::vector<float> positions);
+
+        /**
+         * @brief Set the velocities of each joint in the leg
+         *
+         * @param velocities
+         */
+
```
Comment (ethanmusser) [inline include/kodlab_mjbots_sdk/limb_base.h:137]: Again, the Doxygen comment block should directly precede the function/object/namespace it is documenting.
```suggestion
```

File: include/examples/simple_limb.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
+/**
+ * @file simple_limb.h
+ * @author Zac Gong
+ * @brief A simple LimbBase derived class example, a 1DOF limb with 1 joint at
+ *        the origin and a end effector
+ * @date 2022-09-02
+ *
+ * @copyright Copyright (c) 2022
```
Comment (ethanmusser) [inline include/examples/simple_limb.h:8]:
```suggestion
 * @copyright (c) Copyright 2022 The Trustees of the University of Pennsylvania.
 * All rights reserved.
```

File: include/examples/simple_limb.h [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ */
+
+#pragma once
+
+#include <Eigen/Dense>
+#include "kodlab_mjbots_sdk/limb_base.h"
```
Comment (ethanmusser) [inline include/examples/simple_limb.h:15]: Insert newline between these includes per [this style guide section](https://drake.mit.edu/styleguide/cppguide.html#Names_and_Order_of_Includes).

File: examples/robot_example.cpp [hunk 5]
```diff
@@ -60,16 +63,44 @@ class SimpleRobotControlLoop : public kodlab::mjbots::MjbotsControlLoop<ManyMoto
     }
 };
 
+class LimbTest : public kodlab::LimbBase
```
Comment (ethanmusser) [inline examples/robot_example.cpp:66]: Is this class intended to be committed?  Or is this just for your testing?  If the former, this class definition should live in its own file somewhere with proper commenting.  If the latter, be sure to remove before final merge.

File: include/examples/simple_limb.h [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * @brief A simple 1DOF limb, just 1 joint with a straight end effector with a specified
+ * xyz offset, shows implementation of kinematics
+ *
```
Comment (ethanmusser) [inline include/examples/simple_limb.h:20]:
```suggestion
 * @brief A simple 1-DOF limb consisting of a single joint with a straight 
 * end-effector
 * @details A simple, single-degree-of-freedom limb implementation demonstrating
 * the specification of a Cartesian offset and implementation of kinematics 
 * methods.
```

File: include/examples/simple_limb.h [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+ */
+
+class SimpleLimb : public kodlab::LimbBase
+{
```
Comment (ethanmusser) [inline include/examples/simple_limb.h:24]: Formatting:
```suggestion
class SimpleLimb : public kodlab::LimbBase {
```

File: include/kodlab_mjbots_sdk/limb_base.h [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+ */
+
+#pragma once
+#include <limits>
+#include <memory>
+#include <vector>
+#include <type_traits>
+#include <string>
+#include "kodlab_mjbots_sdk/joint_base.h"
+#include <Eigen/Dense>
```
Comment (ethanmusser) [inline include/kodlab_mjbots_sdk/limb_base.h:19]: Format includes per [this style guide section](https://drake.mit.edu/styleguide/cppguide.html#Names_and_Order_of_Includes).
```suggestion
#pragma once

#include <limits>
#include <memory>
#include <string>
#include <type_traits>
#include <vector>

#include <Eigen/Dense>

#include "kodlab_mjbots_sdk/joint_base.h"
```

File: include/kodlab_mjbots_sdk/limb_base.h [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+namespace kodlab
+{
+    /**
+     * Abstract limb class that takes in a series of joints and treats it as an abstract limb
+     * Contains the kinematics and configurations for a leg
+     * @brief Abstract base class for legs consisting of multiple joints to be used in the robot class
+     */
+
+    struct LimbConfig
```
Comment (ethanmusser) [inline include/kodlab_mjbots_sdk/limb_base.h:29]: Provide Doxygen comment block.

File: include/kodlab_mjbots_sdk/limb_base.h [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    {
+        std::vector<std::array<float, 3>> actuator_offsets; // Offset between centers of two joints, in the frame of the first joint
+        std::vector<Eigen::Matrix3f> actuator_orientations; // Rotation matrices for each joint with respect to the previous joint (begins with R0_1)
+        Eigen::MatrixXf selection_matrix;
```
Comment (ethanmusser) [inline include/kodlab_mjbots_sdk/limb_base.h:33]: Provide Doxygen-friendly comments, e.g., 
```suggestion
        /** 
         * @brief Offset between centers of two joints, in the frame of the 
         * first joint
         */
        std::vector<std::array<float, 3>> actuator_offsets;

        /**
         * @brief Rotation matrix transforming from the previous joint frame to
         * the current joint frame, beginning with \f${}^0 R_1\f$
         */
        std::vector<Eigen::Matrix3f> actuator_orientations;

        /** 
         * @brief Matrix of zeros and ones for re-ordering joints in member 
         * methods
         */
        Eigen::MatrixXf selection_matrix;
```

File: include/kodlab_mjbots_sdk/limb_base.h [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+    class LimbBase
+    {
+    public:
+        /**
+         * @brief Construct a new Limb Base object with a name
+         *
+         * @param name          /// Sets the limb name
+         * @param joints        /// Joints that make up the leg
+         * @param config        /// Configuration of the joints in the leg
```
Comment (ethanmusser) [inline include/kodlab_mjbots_sdk/limb_base.h:44]: The triple-slash here is not correct, see [here](https://doxygen.nl/manual/docblocks.html).
```suggestion
         * @param name Sets the limb name
         * @param joints Joints that make up the leg
         * @param config Configuration of the joints in the leg
```

File: include/kodlab_mjbots_sdk/limb_base.h [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+            std::string name,
+            std::vector<std::shared_ptr<JointBase>> joints,
+            LimbConfig config);
+
+        /**
+         * @brief Construct a new Limb Base object
+         *
+         * @param joints        /// Joints that make up the leg
+         * @param config        /// Configuration of the joints in the leg
+         *
```
Comment (ethanmusser) [inline include/kodlab_mjbots_sdk/limb_base.h:58]: The triple-slash here is not correct, see [here](https://doxygen.nl/manual/docblocks.html).
```suggestion
         * @param joints Joints that make up the leg
         * @param config Configuration of the joints in the leg
         *
```

File: include/kodlab_mjbots_sdk/limb_base.h [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+         * @param name          /// Sets the limb name
+         * @param joints        /// Joints that make up the leg
+         * @param config        /// Configuration of the joints in the leg
+         *
+         */
+
+        LimbBase(
+            std::string name,
+            std::vector<std::shared_ptr<JointBase>> joints,
+            LimbConfig config);
```
Comment (ethanmusser) [inline include/kodlab_mjbots_sdk/limb_base.h:51]: Pass moderate cost-to-move parameters by constant reference (see [ISO C++ reference](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f15-prefer-simple-and-conventional-ways-of-passing-information)).
```suggestion
        LimbBase(const std::string &name,
                 const std::vector<std::shared_ptr<JointBase>> &joints,
                 const LimbConfig &config);
```

File: include/kodlab_mjbots_sdk/limb_base.h [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+         * @brief Construct a new Limb Base object
+         *
+         * @param joints        /// Joints that make up the leg
+         * @param config        /// Configuration of the joints in the leg
+         *
+         */
+
+        LimbBase(
+            std::vector<std::shared_ptr<JointBase>> joints,
+            LimbConfig config);
```
Comment (ethanmusser) [inline include/kodlab_mjbots_sdk/limb_base.h:63]: Pass moderate cost-to-move input parameters by constant reference, e.g., 
```suggestion
        LimbBase(const std::vector<std::shared_ptr<JointBase>> &joints,
                 const LimbConfig &config);
```

File: include/kodlab_mjbots_sdk/limb_base.h [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+        virtual ~LimbBase(){};
+
+        /**
+         * @brief Update position and velocity of each joint within leg
+         *
+         * @param pos_list desired positions of each joint
+         * @param vel_list desired velocities of each joint
+         *
+         */
+        void Update(std::vector<float> pos_list, std::vector<float> vel_list);
```
Comment (ethanmusser) [inline include/kodlab_mjbots_sdk/limb_base.h:74]: Pass moderate cost-to-move input parameters by constant reference (see [here](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f15-prefer-simple-and-conventional-ways-of-passing-information)), i.e., 
```suggestion
        void Update(const std::vector<float> &pos_list, 
                    const std::vector<float> &vel_list);
```

File: include/kodlab_mjbots_sdk/limb_base.h [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+         * measured from the origin of the leg (????)
+         *
+         * @param x_des desired x position
+         * @param y_des desired y position
+         * @param z_des desired z position
+         *
+         * @return std::vector<float> positions
+         */
+
+        virtual void InverseKinematics(std::vector<float> EE_pos) = 0;
```
Comment (ethanmusser) [inline include/kodlab_mjbots_sdk/limb_base.h:101]: Pass by constant reference per [ISO reference](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f15-prefer-simple-and-conventional-ways-of-passing-information):
```suggestion
        virtual void InverseKinematics(const std::vector<float> &EE_pos) = 0;
```

File: include/kodlab_mjbots_sdk/limb_base.h [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+         */
+        std::vector<float> get_torques();
+
+        /**
+         * @brief Set the positions of each joint in the leg
+         *
+         * @param positions
+         */
+
+        void set_positions(std::vector<float> positions);
```
Comment (ethanmusser) [inline include/kodlab_mjbots_sdk/limb_base.h:130]: Pass by constant reference per [ISO reference](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f15-prefer-simple-and-conventional-ways-of-passing-information):
```suggestion
        void set_positions(const std::vector<float> &positions);
```

File: include/kodlab_mjbots_sdk/limb_base.h [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+
+        /**
+         * @brief Set the velocities of each joint in the leg
+         *
+         * @param velocities
+         */
+
+        void set_velocities(std::vector<float> velocities);
```
Comment (ethanmusser) [inline include/kodlab_mjbots_sdk/limb_base.h:138]: Pass by constant reference per [ISO C++ reference](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f15-prefer-simple-and-conventional-ways-of-passing-information):
```suggestion
        void set_velocities(const std::vector<float> &velocities);
```

**Review by ethanmusser — COMMENTED** [review]
Here's some high level things to tackle.  I provided a bunch of inline comments to look at with details.  As you tackle these inline issues, be sure to mark the comments as resolved and cite the fix, if applicable.  

I can dig into the code itself when you are ready for a read-through.  Just re-request my review.

### TODO
- [Commit amend or squash](https://www.atlassian.com/git/tutorials/rewriting-history) commit f1d830c14fb56c0695c072f82dc53ed46f72d882 so that it does not add the `.DS_Store` and `thirdparty/.DS_Store` files.  Not sure what these are but it doesn't have anything to do with the elimb class stuff.
- Everything relating to the Docker container (i.e., commit 6005526ecd1a34838613b4e85a406948e4b1a64d) should be added in its own comprehensive PR, not this one.
- [Fix-up and squash](https://blog.sebastian-daschner.com/entries/git-commit-fixup-autosquash) commit d1de1ff95dd8fe49f648db8247498dba0f39675d to not add the `.clang-format` file.  This should be added to the repository, but I need to customize it first to match our style guide, not the Google style guide that this file follows.  This will be added later in a separate PR.  See #35.
- Take a look at the [ISO C++ Reference](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#main) which contains tons of best-practices.
- Lots of formatting things:
   - There are several formatting changes made to prior code.  This code should be left intact, and not be changed in the scope of this PR.
   - Several [style guide](https://drake.mit.edu/styleguide/cppguide.html) formatting violations.  You should get your auto-formatter configured to at least use the default Google style, perhaps with the attached `.clang-format` file.  Fix the following, plus anything else I mentioned with code comments:
      - All code should use [two-space indents](https://drake.mit.edu/styleguide/cppguide.html#Spaces_vs._Tabs), and have a [hard break at 80 characters](https://drake.mit.edu/styleguide/cppguide.html#Line_Length).  The only exception to the two-space intents are modifications to existing code which already uses four-space indents.
      - Opening brackets should [follow on the same line as a function/object/namespace declaration](https://drake.mit.edu/styleguide/cppguide.html#Function_Declarations_and_Definitions), e.g., 
      ```cpp
      void ImproperlyFormattedFunction()
      {
        return;
      }
      
      void ProperlyFormattedFunction() {
        return;
      }
      ```
      - Namespace contents should not be indented, per [this section](https://drake.mit.edu/styleguide/cppguide.html#Namespace_Formatting).
   - All functions, objects, and object members should have a [Doxygen-friendly comment block](https://doxygen.nl/manual/docblocks.html) associated with them for documentation generation.

**Review by ethanmusser — COMMENTED** [review]
I didn't mean to post this but can't remove it...

File: include/examples/simple_limb.h (hunk 4 above)
Comment (zacgong) [reply include/examples/simple_limb.h:15]: done

File: include/kodlab_mjbots_sdk/limb_base.h (hunk 2 above)
Comment (zacgong) [reply include/kodlab_mjbots_sdk/limb_base.h:137]: fixed for one larger formatting commit

File: include/examples/simple_limb.h (hunk 3 above)
Comment (zacgong) [reply include/examples/simple_limb.h:8]: fixed for one larger formatting commit

File: include/examples/simple_limb.h (hunk 6 above)
Comment (zacgong) [reply include/examples/simple_limb.h:20]: fixed for one larger formatting commit

File: include/kodlab_mjbots_sdk/limb_base.h (hunk 8 above)
Comment (zacgong) [reply include/kodlab_mjbots_sdk/limb_base.h:19]: fixed for larger formatting commit

File: include/examples/simple_limb.h (hunk 7 above)
Comment (zacgong) [reply include/examples/simple_limb.h:24]: fixed for larger formatting commit

File: include/kodlab_mjbots_sdk/limb_base.h (hunk 9 above)
Comment (zacgong) [reply include/kodlab_mjbots_sdk/limb_base.h:29]: fixed for larger formatting commit

File: include/kodlab_mjbots_sdk/limb_base.h (hunk 10 above)
Comment (zacgong) [reply include/kodlab_mjbots_sdk/limb_base.h:33]: fixed for larger formatting commit

File: include/kodlab_mjbots_sdk/limb_base.h (hunk 11 above)
Comment (zacgong) [reply include/kodlab_mjbots_sdk/limb_base.h:44]: fixed for larger formatting commit

File: include/kodlab_mjbots_sdk/limb_base.h (hunk 12 above)
Comment (zacgong) [reply include/kodlab_mjbots_sdk/limb_base.h:58]: fixed for larger formatting commit

File: include/kodlab_mjbots_sdk/limb_base.h (hunk 13 above)
Comment (zacgong) [reply include/kodlab_mjbots_sdk/limb_base.h:51]: fixed for larger formatting commit

File: include/kodlab_mjbots_sdk/limb_base.h (hunk 14 above)
Comment (zacgong) [reply include/kodlab_mjbots_sdk/limb_base.h:63]: fixed for larger formatting commit

File: include/kodlab_mjbots_sdk/limb_base.h (hunk 15 above)
Comment (zacgong) [reply include/kodlab_mjbots_sdk/limb_base.h:74]: fixed for larger formatting commit

File: include/kodlab_mjbots_sdk/limb_base.h (hunk 16 above)
Comment (zacgong) [reply include/kodlab_mjbots_sdk/limb_base.h:101]: fixed for larger formatting commit

File: include/kodlab_mjbots_sdk/limb_base.h (hunk 17 above)
Comment (zacgong) [reply include/kodlab_mjbots_sdk/limb_base.h:130]: fixed for larger formatting commit

File: include/kodlab_mjbots_sdk/limb_base.h (hunk 18 above)
Comment (zacgong) [reply include/kodlab_mjbots_sdk/limb_base.h:138]: fixed for larger formatting commit

**Comment by ethanmusser** [comment]
Something seems to have gone wrong with your Git history causing commits in `master` to be overridden.  Closing this PR because it would improperly modify commit history.  Development on this feature has moved to #67.

**Pull request closed without merging**